relay-server/src/actors/project_cache.rs

```diff
@@ -400,6 +372,83 @@ struct UpdateProjectState {
     no_cache: bool,
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+struct QueueKey {
+    key: ProjectKey,
+    sampling_key: ProjectKey,
+}
+
+impl QueueKey {
+    fn new(key: ProjectKey, sampling_key: ProjectKey) -> Self {
+        Self { key, sampling_key }
+    }
+}
+
+/// The queue (buffer) of the incoming envelopes.
+#[derive(Debug, Default)]
+struct Queue {
+    /// Contains the cache of the incoming envelopes.
+    buffer: BTreeMap<QueueKey, Vec<(Box<Envelope>, EnvelopeContext)>>,
+    /// Index of the buffered project keys.
+    index: BTreeMap<ProjectKey, BTreeSet<QueueKey>>,
+}
+
+impl Queue {
+    /// Creates an empty queue.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Adds the value to the queue for the provided key.
+    pub fn enqueue(&mut self, key: QueueKey, value: (Box<Envelope>, EnvelopeContext)) {
+        self.index.entry(key.key).or_default().insert(key);
+        self.index.entry(key.sampling_key).or_default().insert(key);
```

Comment:
I don't understand the purpose of another entry with `key.sampling_key`, why are we doing that?